### ISSUE TITLE
[regression] extra option -ffast-math bring in inefficient assembly output

### ISSUE THREAD

**Issue by vfdff** [llvm/llvm-project#64244]
test （simplified from #62820）: https://gcc.godbolt.org/z/jv4Kzj9rY
```
double **vatom;

void ev_tally(int i, int j, int nlocal, int newton_pair,
        double evdwl, double ecoul, double fpair,
        double delx, double dely, double delz)
{
  double evdwlhalf,ecoulhalf,epairhalf,v[6];

  v[0] = delx*delx*fpair;
  v[1] = dely*dely*fpair;
  v[2] = delz*delz*fpair;
  v[3] = delx*dely*fpair;
  v[4] = delx*delz*fpair;
  v[5] = dely*delz*fpair;


    vatom[i][0] += 0.5*v[0];
    vatom[i][1] += 0.5*v[1];
    vatom[i][2] += 0.5*v[2];
    vatom[i][3] += 0.5*v[3];
    vatom[i][4] += 0.5*v[4];
    vatom[i][5] += 0.5*v[5];
}
```
Now, the clang enable `-ffp-contract=on` default, and it have efficient assembly output.
Then, I try with extra option  `-ffast-math`, it has some regression.

**Comment by vfdff** [comment]
Discussion on https://discourse.llvm.org/t/fp-contraction-fma-on-by-default/64975
 - The behavior controlled by -ffp-contract=on is compliant with the C standard,  double fma(double a, double b) { return a * b + 5; } can emit an fusedMultiplyAdd operation
 - The behavior controlled by -ffp-contract=fast enables non-C-standard-compliant behavior that ignores the #pragma and will create contractions after other optimizations like inlining, and **across expressions**

**Comment by vfdff** [comment]
candidate MR:  https://reviews.llvm.org/D156653

**Comment by arsenm** [comment]
I don't think there's a representational issue to solve, I think there's a backend combine to fix

**Comment by vfdff** [comment]
Yes, the issue also disappear with  **-ffast-math -fno-associative-math**, https://gcc.godbolt.org/z/TcqE8z6qP
